src/v/serde/protobuf/parser.cc

```diff
@@ -0,0 +1,522 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "serde/protobuf/parser.h"
+
+#include "bytes/iobuf_parser.h"
+#include "utils/vint.h"
+
+#include <google/protobuf/descriptor.h>
+
+#include <algorithm>
+#include <iterator>
+#include <limits>
+#include <memory>
+#include <stdexcept>
+#include <type_traits>
+#include <variant>
+
+namespace serde::pb {
+
+namespace pb = google::protobuf;
+
+enum class wire_type : uint8_t {
+    variant = 0,
+    i64 = 1,
+    length = 2,
+    group_start = 3,
+    group_end = 4,
+    i32 = 5,
+    max = 5,
+};
+
+struct tag {
+    wire_type wire_type;
+    int32_t field_number;
+};
+
+int32_t zig_zag_decode(uint32_t n) {
+    // Note:  Using unsigned types prevent undefined behavior
+    return static_cast<int32_t>((n >> 1U) ^ (~(n & 1U) + 1));
+}
+
+template<typename T>
+T read_varint(iobuf_parser_base& parser) {
+    constexpr size_t limit = sizeof(T) == 8 ? 10 : 5;
+    auto buf = parser.peek_bytes(std::min(limit, parser.bytes_left()));
+    if constexpr (std::is_same_v<T, uint32_t>) {
+        auto [v, len] = unsigned_vint::detail::deserialize(buf, limit);
+        parser.skip(len);
+        return static_cast<uint32_t>(v);
+    } else if constexpr (std::is_same_v<T, int32_t>) {
+        auto [v, len] = unsigned_vint::detail::deserialize(buf, limit);
+        parser.skip(len);
+        return zig_zag_decode(static_cast<uint32_t>(v));
+    } else if constexpr (std::is_same_v<T, uint64_t>) {
+        auto [v, len] = unsigned_vint::detail::deserialize(buf, limit);
+        parser.skip(len);
+        return v;
+    } else {
+        static_assert(std::is_same_v<T, int64_t>);
+        auto [v, len] = unsigned_vint::detail::deserialize(buf, limit);
+        parser.skip(len);
+        return vint::decode_zigzag(v);
+    }
+}
```

Comment:
iobuf depends on vint so we can't do that. I think this makes sense to keep here, especially in light of your other comments this might change slightly

Reply:
Thoughts now?

Reply:
This is very clear now, thank you! I was going to ask why not put in iobuf_parser_base like read_unsigned_varint but the comments make it clear that the zigzag handling is tied to the pb wire protocol